worker/transfer.go

```diff
@@ -125,46 +128,107 @@ func parallelUploadSlab(ctx context.Context, sp storeProvider, shards [][]byte,
 		span.End()
 	}
 
-	// spawn workers and send initial requests
-	hostIndex := 0
+	// helper to track used hosts.
+	hostUsed := make([]bool, len(contracts))
+	nextHost := func() int {
+		for i := range hostUsed {
+			if !hostUsed[i] {
+				hostUsed[i] = true
+				return i
+			}
+		}
+		return -1
+	}
+
+	// helper to determine whether a request is finished.
+	isReqFinished := func(r req) bool {
+		select {
+		case <-r.finishedCtx.Done():
+			return true
+		default:
+			return false
+		}
+	}
+
+	// helper function for launching worker given an existing request. Returns
+	// true if a worker was launched for the request or if the request was
+	// already finished.
 	inflight := 0
-	for i := range shards {
-		go worker(req{contracts[hostIndex], i})
-		hostIndex++
+	launchWorker := func(r req) bool {
+		if isReqFinished(r) {
+			return true
+		}
+		hostIndex := nextHost()
+		if hostIndex == -1 {
+			return false
+		}
+		go worker(req{r.finishedCtx, r.finishedFn, hostIndex, r.shardIndex})
 		inflight++
+		return true
+	}
+
+	// spawn workers and send initial requests
+	for i := range shards {
+		finishedCtx, finishedFn := context.WithCancel(ctx)
+		if !launchWorker(req{finishedCtx, finishedFn, -1, i}) {
+			panic("failed to launch worker for initial shard - should never happen")
+		}
 	}
 
 	// collect responses
 	var errs HostErrorSet
 	sectors := make([]object.Sector, len(shards))
 	rem := len(shards)
+	var neededOverdrive []req
 	for rem > 0 && inflight > 0 {
```

Comment:
I wonder if it would be better if we would change this to `for inflight > 0 {` and have a `break` right after `rem--` if it's zero. I guess that's a `nit` but I think it's better because it says "as long as there's a request in flight, wait for a response and then handle it"